watershed.py

```diff
@@ -0,0 +1,334 @@
+    image_raveled, marker_coords, offsets, mask, output, strides = prepped_data
+    output = slow_raveled_watershed(image_raveled, marker_coords, 
+                                    offsets, mask, strides, compactness, 
+                                    output, affinities)
+    if affinities:
+        shape = image.shape[1:]
+    else:
+        shape = image.shape
+    output = output.reshape(shape)
+    return output
+
+
+def prep_data(image, marker_coords, mask=None, affinities=False, output=None):
+    # INTENSITY VALUES
+    if affinities:
+        im_ndim = image.ndim - 1 # the first dim should represent affinities
+        image_shape = image.shape[1:]
+        image_strides = image[0].strides
+        image_itemsize = image[0].itemsize
+        raveled_image = np.zeros((image.shape[0], image[0].size), dtype=image.dtype)
+        for i in range(image.shape[0]):
+            raveled_image[i] = image[i].ravel()
+    else:
+        im_ndim = image.ndim
+        image_shape = image.shape
+        image_strides = image.strides
+        image_itemsize = image.itemsize
+        raveled_image = image.ravel()
+    # NEIGHBORS
+    selem, centre = _validate_connectivity(im_ndim, 1, None)
+    if affinities:
+        # array of shape (ndim * 2, 2) giving the indicies of neighbor affinities
+        offsets = _indices_to_raveled_affinities(image_shape, selem, centre)
+    else:
+        offsets = _offsets_to_raveled_neighbors(image_shape, selem, centre)
+    raveled_markers = np.apply_along_axis(raveled_coordinate, 1, 
+                                          marker_coords, **{'shape':image_shape})
+    if mask is None:
+        small_shape = [s - 2 for s in image_shape]
+        mask = np.ones(small_shape, dtype=bool)
+        mask = np.pad(mask, 1, constant_values=0)
+        assert image_shape == mask.shape
+    mask_raveled = mask.ravel()
+    if output is None:
+        output = np.zeros(mask_raveled.shape, dtype=raveled_image.dtype)
+        labels = np.arange(len(raveled_markers)) + 1
+        output[raveled_markers] = labels
+    strides = np.array(image_strides, dtype=np.intp) // image_itemsize
+    return raveled_image, raveled_markers, offsets, mask_raveled, output, strides
+
+
+def raveled_coordinate(coordinate, shape):
+    # array[z, y, x] = array.ravel()[z * array.shape[1] * array.shape[2] + y * array.shape[2] + x]
+    raveled_coord = 0
+    for i in range(len(coordinate)):
+        to_add = coordinate[i]
+        for j in range(len(shape)):
+            if j > i:
+                to_add *= shape[j]
+        raveled_coord += to_add
+    return raveled_coord
+    
+
+def _indices_to_raveled_affinities(image_shape, selem, centre):
+    im_offsets = _offsets_to_raveled_neighbors(image_shape, selem, centre)
+    affs = np.concatenate([np.arange(len(image_shape)), 
+                           np.arange(len(image_shape))[::-1]])
+    indices = np.stack([affs, im_offsets], axis=1)
```

Comment:
I'm super confused here. How sure are you about this? I think the `im_offsets` should be *0* for the -1 offsets, and equal to the neighbors offset for the +1 offsets... In other words I think you have an off-by-one error here on the -1 side.

*if* `_offsets_to_raveled_neighbors` is sorted, then you're *almost* correct, you just need to zero out the first half of `im_offsets`.